src/analysis/convex/cone.lean

```diff
@@ -707,4 +737,93 @@ begin
+
+  -- f is continuous at 0 from the right
+  have fc : continuous_within_at f (set.Ioi (0 : ℝ)) 0 :=
+    (continuous_id.smul continuous_const).continuous_within_at,
+
+  -- 0 belongs to the closure of the f (0, ∞)
+  have mem₀ := fc.mem_closure_image (by rw [closure_Ioi (0 : ℝ), set.mem_Ici]),
+  have f₀ : f 0 = 0 := zero_smul ℝ x,
+  rw f₀ at mem₀,
+
+  -- as 0 ∈ closure f (0, ∞) and closure f (0, ∞) ⊆ K, 0 ∈ K.
+  rw [convex_cone.pointed, ← set_like.mem_coe],
+  exact set.mem_of_subset_of_mem clf mem₀,
+end
+
+section complete_space
+variables [complete_space H]
+
+/-- This is a stronger version of the Hahn-Banach separation theorem for closed convex cones. This
+is also the geometric interpretation of Farkas' lemma. -/
+theorem hyperplane_separation_point_nonempty_closed_convex_cone {K : convex_cone ℝ H}
+  (ne : (K : set H).nonempty) (hc : is_closed (K : set H)) {b : H} (disj : b ∉ K) :
+  ∃ (y : H), (∀ x : H, x ∈ K → 0 ≤ ⟪x, y⟫_ℝ) ∧ ⟪y, b⟫_ℝ < 0 :=
+begin
+  -- let `z` be the point in `K` closest to `b`
+  obtain ⟨z, hzK, infi⟩ := exists_norm_eq_infi_of_complete_convex ne hc.is_complete K.convex b,
+
+  -- for any `w` in `K`, we have `⟪b - z, w - z⟫_ℝ ≤ 0`
+  have hinner := (norm_eq_infi_iff_real_inner_le_zero K.convex hzK).1 infi,
+
+  -- set `y := z - b`
+  use z - b,
+
+  split,
+  begin
+  -- the rest of the proof is a straightforward calculation
+    rintros x hxK,
+    specialize hinner _ (K.add_mem hxK hzK),
+    rwa [add_sub_cancel, real_inner_comm, ← neg_nonneg, neg_eq_neg_one_mul,
+         ← real_inner_smul_right, neg_smul, one_smul, neg_sub] at hinner,
+  end,
+  begin
+    -- as `K` is closed and non-empty, it is pointed
+    have hinner₀ := hinner 0 (pointed_of_nonempty_closed ne hc),
+
+    -- the rest of the proof is a straightforward calculation
+    rw [zero_sub, inner_neg_right, right.neg_nonpos_iff] at hinner₀,
+    have hbz : b - z ≠ 0 := by { rw sub_ne_zero, contrapose! hzK, rwa ← hzK },
+    rw [← neg_zero, lt_neg, ← neg_one_mul, ← real_inner_smul_left, smul_sub, neg_smul, one_smul,
+      neg_smul, neg_sub_neg, one_smul],
+    calc 0 < ⟪b - z, b - z⟫_ℝ : lt_of_not_le ((iff.not real_inner_self_nonpos).2 hbz)
+    ... = ⟪b - z, b - z⟫_ℝ + 0 : (add_zero _).symm
+    ... ≤ ⟪b - z, b - z⟫_ℝ + ⟪b - z, z⟫_ℝ : add_le_add rfl.ge hinner₀
+    ... = ⟪b - z, b - z + z⟫_ℝ : inner_add_right.symm
+    ... = ⟪b - z, b⟫_ℝ : by rw sub_add_cancel,
+  end,
+end
+
+/-- The inner dual of inner dual of a non-empty, closed convex cone is itself.  -/
+theorem inner_dual_cone_of_inner_dual_cone_eq_self
+  {K : convex_cone ℝ H} (ne : (K : set H).nonempty) (hc : is_closed (K : set H)) :
+  ((K : set H).inner_dual_cone : set H).inner_dual_cone = K := convex_cone.ext $ λ x, iff.intro
+begin
+  rw [mem_inner_dual_cone, ← set_like.mem_coe],
+  contrapose!,
+  rintro hx,
+  exact hyperplane_separation_point_nonempty_closed_convex_cone ne hc hx,
+end
+begin
+  rintro hxK y h,
+  specialize h x hxK,
+  rwa real_inner_comm,
+end
```

Comment:
```suggestion
  ((K : set H).inner_dual_cone : set H).inner_dual_cone = K :=
begin
  ext x,
  split,
  { rw [mem_inner_dual_cone, ← set_like.mem_coe],
    contrapose!,
    exact hyperplane_separation_point_nonempty_closed_convex_cone ne hc },
  { rintro hxK y h,
    specialize h x hxK,
    rwa real_inner_comm },
end
```
This is the more common style in mathlib, also `ext` is very handy tactic and most of the time better than calling the corresponding ext-lemma yourself. `intro x, exact foo x` can almost always be replaced by `exact foo`.

Reply:
in the second block you could also write
```
rw real_inner_comm,
exact h x hxK
```
which feels more natural to me, but I have really no strong feelings on that.

Reply:
Thanks!